### PR TITLE
Support base64 and base32768 file name encoding for crypt.

### PR DESCRIPTION
**Pull request by tinytangent**
#### What is the purpose of this change?

This PR add support for base64 and base32768 file name encoding for crypt. This resolves https://github.com/rclone/rclone/issues/5801 and relieve the problem of file name length being too long when encrypted.

It is based on the commit 78cab2ba2071dd4c58a34afa22486ad1dec306fc by @Max-Sum . For other changes in https://github.com/Max-Sum/rclone/commits/compress_crypted_filename, I' m worried that some of them might break backward compatibility or involving security issue, and I' m not very familiar in cryptography.

#### Was the change discussed in an issue or in the forum before?

https://forum.rclone.org/t/whats-going-on-with-the-proposal-about-base32768-file-name-encoding/27341
https://forum.rclone.org/t/base32768-to-compress-filename-length/13202

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/crypt/cipher.go
Comment (ncw) [inline backend/crypt/cipher.go:117]: This comment didn't want to get deleted!

File: backend/crypt/cipher.go
Comment (ncw) [inline backend/crypt/cipher.go:126]: This seems over complicated to me....

I'd suggest you define this (note lower case so we don't make a public symbol)

```go
type caseInsensitiveBase32Encoding struct {}
```

Then define the two methods

```
func (_ caseInsensitiveBase32Encoding) EncodeToString(src []byte) string {
// contents the same as encodeFileName in original source
}
func (_ caseInsensitiveBase32Encoding) DecodeString(s string) ([]byte, error) {
// contents the same as decodeFileName in original source
}
```

And finally check it satisfies the interface

```
var _ StrEncoding = caseInsensitiveBase32Encoding{}
```

File: backend/crypt/cipher.go
Comment (ncw) [inline backend/crypt/cipher.go:121]: This should be called `filenameEncoding` - initial lower so it isn't exported and more specific name

File: backend/crypt/cipher.go
Comment (ncw) [inline backend/crypt/cipher.go:149]: ...this then becomes

    enc = caseInsensitiveBase32Encoding{}

**Review by ncw — COMMENTED** [review]
I put some initial comments inline.

This should probably also have some tests in `backend/crypt/crypt_test.go` to run the backend tests against the new file name encoding.

I'm going to allow the CI to run now - expect some linting problems you'll need to fix up!

File: backend/crypt/cipher.go
Comment (tinytangent) [reply backend/crypt/cipher.go:117]: Sorry about this. Fixed in 79817be

File: backend/crypt/cipher.go
Comment (tinytangent) [reply backend/crypt/cipher.go]: Indeed this change is modifying codes that already work. I reverted the base32 implementation in 79817be.

**Comment by tinytangent** [comment]
I also added some new fstests in `backend/crypt/crypt_test.go` for the newly added base64 and base32768 encoding in 046656d.

**Comment by tinytangent** [comment]
Linting problems should now be fixed in fb0fb43.

**Review by ncw — COMMENTED** [review]
This looks great now - thank you.

Can you rebase it on master, fix the conflicts and squash it down to two commits - one for Max-Sum's work and one for your work.

(can you make sure Max-Sum commit compiles on its own - if it doesn't fix it!)

I'll merge it then - thank you.

PS If any of this git stuff is too much for you then I can do it easily enough.

**Comment by tinytangent** [comment]
Thank you~

I have reorganized the commit history to make it cleaner. So now d2e180f contains all the code changes including go.sum and go.mod, and it compiles on it own  (but will fail most of test cases due to internal interfaces changes).

1bd9007 updates test cases and docs.

**Comment by tinytangent** [comment]
There is another thing to note: Currently there is a failing test case in crypt, but it is unlikely introduced by my changes because it is also happening on the master branch, it looks like this:

```
--- FAIL: TestNoDataObfuscate (0.53s)
    fstests.go:418: Using remote "TestCrypt4:"
    --- FAIL: TestNoDataObfuscate/FsMkdir (0.49s)
        --- FAIL: TestNoDataObfuscate/FsMkdir/FsPutFiles (0.39s)
            --- FAIL: TestNoDataObfuscate/FsMkdir/FsPutFiles/Internal (0.00s)
                --- FAIL: TestNoDataObfuscate/FsMkdir/FsPutFiles/Internal/ObjectInfo (0.00s)
                    crypt_internal_test.go:94: 
                                Error Trace:    crypt_internal_test.go:94
                                                                        crypt_internal_test.go:140
                                Error:          Not equal: 
                                                expected: 148
                                                actual  : 100
                                Test:           TestNoDataObfuscate/FsMkdir/FsPutFiles/Internal/ObjectInfo
                --- FAIL: TestNoDataObfuscate/FsMkdir/FsPutFiles/Internal/ObjectInfoWrap (0.00s)
                    crypt_internal_test.go:94: 
                                Error Trace:    crypt_internal_test.go:94
                                                                        crypt_internal_test.go:141
                                Error:          Not equal: 
                                                expected: 148
                                                actual  : 100
                                Test:           TestNoDataObfuscate/FsMkdir/FsPutFiles/Internal/ObjectInfoWrap
```

I' m not very certain if such a failure is expected or something went wrong. Maybe I should open another issue about this?

**Review by ncw — APPROVED** [review]
That looks great now - thank you - I will merge now.

**Comment by ncw** [comment]
...actually I will wait for the CI to complete before merging!

**Comment by ncw** [comment]
> There is another thing to note: Currently there is a failing test case in `backend/crypt`, but it is unlikely introduced by my changes because it is also happening on the `master` branch, it looks like this:
[snip]
> I' m not very certain if such a failure is expected or something went wrong. Maybe I should open another issue about this?

That is very interesting - thank you.

What that shows up is that test isn't being run on the CI, nor in the integration tests which is a bit of an oversight, so thank you for bringing it to my attention.

I'll fix the test not being run and the test itself shortly :-)

**Comment by Max-Sum** [comment]
Onedrive used to have a bug on processing special characters. Microsft ignores this issue and therefor I didn't continue on developing this. Sorry for that.
You could test if it works on Onedrive now.

**Pull request merged**
